internal/queue/loops/counters/counters.go

```diff
@@ -0,0 +1,241 @@
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync/atomic"
+
+	"github.com/dapr/kit/ptr"
+
+	"github.com/diagridio/go-etcd-cron/api"
+	"github.com/diagridio/go-etcd-cron/internal/api/queue"
+	"github.com/diagridio/go-etcd-cron/internal/counter"
+	"github.com/diagridio/go-etcd-cron/internal/queue/actioner"
+	"github.com/diagridio/go-etcd-cron/internal/queue/loops"
+)
+
+type Options struct {
+	IDx      *atomic.Int64
+	Actioner actioner.Interface
+	Name     string
+}
+
+type counters struct {
+	act  actioner.Interface
+	name string
+
+	idx     *atomic.Int64
+	cancel  context.CancelFunc
+	counter counter.Interface
+}
+
+func New(opts Options) loops.Interface[*queue.JobAction] {
+	return loops.New(loops.Options[*queue.JobAction]{
+		Handler: &counters{
+			act:  opts.Actioner,
+			name: opts.Name,
+			idx:  opts.IDx,
+		},
+		BufferSize: ptr.Of(uint64(5)),
+	})
+}
+
+func (c *counters) Handle(ctx context.Context, event *queue.JobAction) error {
+	switch action := event.GetAction().(type) {
+	case *queue.JobAction_Informed:
+		return c.handleInformed(ctx, action.Informed)
+
+	case *queue.JobAction_ExecuteRequest:
+		return c.handleExecuteRequest(ctx, action.ExecuteRequest)
+
+	case *queue.JobAction_ExecuteResponse:
+		return c.handleExecuteResponse(ctx, action.ExecuteResponse)
+
+	case *queue.JobAction_Deliverable:
+		c.handleDeliverable()
+		return nil
+
+	case *queue.JobAction_Close:
+		c.handleClose()
+		return nil
+
+	default:
+		return fmt.Errorf("unknown inner control event action: %T", action)
+	}
+}
+
+func (c *counters) handleInformed(ctx context.Context, action *queue.Informed) error {
+	if c.cancel != nil {
+		c.cancel()
+		c.cancel = nil
+	}
+
+	c.act.Unstage(c.name)
+
+	if action.GetIsPut() {
+		c.idx.Store(action.GetJobModRevision())
+
+		var err error
+		c.counter, err = c.act.Schedule(ctx, c.name, action.GetJobModRevision(), action.GetJob())
+		return err
+	}
+
+	// Delete counter
+	if c.counter != nil {
+		c.act.Deschedule(c.counter)
+		c.close()
+	}
+
+	return nil
+}
+
+func (c *counters) handleExecuteRequest(ctx context.Context, action *queue.ExecuteRequest) error {
+	counter := c.counter
+	if counter == nil {
+		return errors.New("catastrophic state machine error: lost counter")
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	doneCh := make(chan struct{})
+
+	c.cancel = func() { cancel(); <-doneCh }
+
+	idx := c.idx.Load()
+
+	go func() {
+		defer func() {
+			cancel()
+			close(doneCh)
+		}()
+
+		result := c.act.Trigger(ctx, counter.TriggerRequest())
+
+		c.act.AddToControlLoop(&queue.ControlEvent{
+			Action: &queue.ControlEvent_ExecuteResponse{
+				ExecuteResponse: &queue.ExecuteResponse{
+					JobName:    action.GetJobName(),
+					CounterKey: action.GetCounterKey(),
+					Result:     result,
+					Uid:        idx,
+				},
+			},
+		})
+	}()
+
+	return nil
+}
+
+func (c *counters) handleExecuteResponse(ctx context.Context, action *queue.ExecuteResponse) error {
+	// Ignore if the execution response jf the idx has been changed.
```

Comment:
nit: if

edit: looks like a lot of `i`s are `j`s in this file

Reply:
Thanks. sed replace error.